hydra_base/lib/attributes.py

```diff
@@ -1085,3 +1109,249 @@ def delete_attribute_group_items(attributegroupitems, **kwargs):
+                if ra_rs is None:
+                    log.info("A duplicate found for %s on %s %s. Deleting as it has no data.",
+                             ra_to_delete.attr.name,
+                             ra_to_delete.ref_key,
+                             resource.name)
+                    db.DBSession.delete(ra_to_delete)
+                else:
+                    log.info("A duplicate found for %s on %s %s. Not deleting as it has data.",
+                             ra_to_delete.attr.name,
+                             ra_to_delete.ref_key,
+                             resource.name)
+
+            #no need to go further, so continue the outer loop
+            continue
+
+        #do a second pass, this time transferring data from the bad resourceattributes
+        #to the keepers
+        for duplicate_ra in duplicate_ras:
+            #focus now on the keepers i.e. the ones defined in the template
+            if duplicate_ra.attr_id in resource_typeattrs:
+                ra_rs = db.DBSession.query(ResourceScenario)\
+                    .filter(ResourceScenario.resource_attr_id == duplicate_ra.id).first()
+
+                dupes_rs = data_to_transfer.get(ra.attr.name)
+                if ra_rs is None:
+                    #if this has no value, see if the dupe has a value and transfer it
+                    if dupes_rs is not None:
+                        log.info("Updating resource scenario with new reference to %s",
+                                 duplicate_ra.attr.name)
+                        newrs = ResourceScenario()
+                        newrs.scenario_id = dupes_rs.scenario_id
+                        newrs.resource_attr_id = duplicate_ra.id
+                        newrs.dataset_id = dupes_rs.dataset_id
+                        db.DBSession.add(newrs)
+                else:
+                    #delete any RS which are not needed
+                    if dupes_rs is not None:
+                        db.DBSession.delete(dupes_rs)
+
+
+        #now finally delete all the dupes
+        for ra_to_delete in ras_to_delete:
+
+            log.info("Deleting resource attr %s (id: %s) from %s %s",\
+                     duplicate_ra.attr.name,\
+                     duplicate_ra.id,\
+                     duplicate_ra.ref_key,\
+                     resource.name)
+            db.DBSession.delete(ra_to_delete)
+
+
+def delete_duplicate_attributes(dupe_list):
+    """
+        Take a list of duplicate attributes and delete all but one.
+        Steps are:
+            1: Select one of the duplicates to keep
+            2: Remap all resource attributes and type attributes to point
+               from duplicate attrs to the keeper
+            3: Delete the duplicates
+        args:
+            dupe_list: a list of iterables containing duplicate attributes
+    """
+
+    #sort by ID, and choose the attribute with the smallest ID to keep
+    #The rationale being that this likely has the most existing references
```

Comment:
Good question. It would require querying the tResourceAttr and tTypeAttr table for each attr_id and counting the combined links for each one, then choosing the smallest number.

Given that this has no functional bearing on the process (they all end up pointing to one place) I don't think it's worth that extra degree of complexity.

Reply:
Yes, that sounds expensive given that 'most references' is just as arbitrary as 'lowest id' for most purposes.

For more explicitness, the func could return the `keeper` (1283), so callers would be informed explicitly to which attr the references were remapped.